expression parser #2

fkleuver opened the pull request:
Fixes #200

AureliaEffect commented:
This pull request **introduces 1 alert** and **fixes 2** when merging b0fd2c32d14f7e09ca9013468e1aa9f2ac850854 into a0c9b0ecd41c556f9ea20f409f94810c8eb5a3de - [view on LGTM.com](https://lgtm.com/projects/g/aurelia/aurelia/rev/pr-f04d9fd8ffcadadc7ec4d3981e0df2147ed074f1)

**new alerts:**

* 1 for Unused variable, import, function or class

**fixed alerts:**

* 2 for Unused variable, import, function or class

---

*Comment posted by [LGTM.com](https://lgtm.com)*

AureliaEffect commented:
This pull request **introduces 1 alert** and **fixes 2** when merging 77aabc77d70e86b3645b949a18ae96259b0dc88a into a0c9b0ecd41c556f9ea20f409f94810c8eb5a3de - [view on LGTM.com](https://lgtm.com/projects/g/aurelia/aurelia/rev/pr-4e5555bbbb3de3b95bac2275c018eafe3637d8dc)

**new alerts:**

* 1 for Unused variable, import, function or class

**fixed alerts:**

* 2 for Unused variable, import, function or class

---

*Comment posted by [LGTM.com](https://lgtm.com)*

AureliaEffect commented:
This pull request **introduces 1 alert** and **fixes 2** when merging 38760bbeef8c21739134c0bed045663373a947db into a0c9b0ecd41c556f9ea20f409f94810c8eb5a3de - [view on LGTM.com](https://lgtm.com/projects/g/aurelia/aurelia/rev/pr-b6dd0aae120bc716f77be3059226958fd5a8235e)

**new alerts:**

* 1 for Unused variable, import, function or class

**fixed alerts:**

* 2 for Unused variable, import, function or class

---

*Comment posted by [LGTM.com](https://lgtm.com)*

AureliaEffect commented:
This pull request **introduces 2 alerts** and **fixes 4** when merging 3c6627c38662641c4210ef553ff34888d473efcb into a0c9b0ecd41c556f9ea20f409f94810c8eb5a3de - [view on LGTM.com](https://lgtm.com/projects/g/aurelia/aurelia/rev/pr-d0dd6180744b0f598a8b38d3444f76c919714c65)

**new alerts:**

* 2 for Unused variable, import, function or class

**fixed alerts:**

* 4 for Unused variable, import, function or class

---

*Comment posted by [LGTM.com](https://lgtm.com)*

AureliaEffect commented:
This pull request **introduces 2 alerts** and **fixes 4** when merging 4565244b6d181154b5ad481c8a2c57a5c9f1b320 into a0c9b0ecd41c556f9ea20f409f94810c8eb5a3de - [view on LGTM.com](https://lgtm.com/projects/g/aurelia/aurelia/rev/pr-6569010298ac6c13611e794752ccaaab710e5b00)

**new alerts:**

* 2 for Unused variable, import, function or class

**fixed alerts:**

* 4 for Unused variable, import, function or class

---

*Comment posted by [LGTM.com](https://lgtm.com)*

AureliaEffect commented:
This pull request **introduces 2 alerts** and **fixes 4** when merging 4b0e45f2d9f5e1d5a87aadddcec49846297816ce into a0c9b0ecd41c556f9ea20f409f94810c8eb5a3de - [view on LGTM.com](https://lgtm.com/projects/g/aurelia/aurelia/rev/pr-6047a1c14c35af70d9d78eff859fa08b4b86ac1f)

**new alerts:**

* 2 for Unused variable, import, function or class

**fixed alerts:**

* 4 for Unused variable, import, function or class

---

*Comment posted by [LGTM.com](https://lgtm.com)*

fkleuver commented:
@EisenbergEffect @bigopon This piece is ready for review/merge. In summary:

### ExpressionParser
- ForOfStatement and Interpolation are now also fully tested, bringing the parser to 100% coverage (with a test density that gives confidence)
- The caching issue with the runtime expression parser is fixed. There are now 3 separate lookups corresponding to the 3 main BindingTypes: Interpolation, ForOfCommand, and the rest. These caches can no longer return parsed expressions for the wrong type.
- parse/parseCore return types are strongly typed to return the correct set of possible expressions depending on bindingType passed in

### @decoratable
- Re-introduced the `@connectable` decorator now that there is more than one binding type that can use it. Refactored `Binding` and `LetBinding` to use the decorator.
- Introducated `Decoratable` and `Decorated` interfaces in the kernel that make it easier to write strongly typed decorators. `@connectable` uses these and provides two overloads (invocation and non-invocation), both of which will type-check the class it is applied to and give a compile time type error if the class doesn't implement the correct interface. Should apply this to other decorators as well.

### Interpolation
- Implemented the `InterpolationBinding` based on the one in `templating-binding` from vCurrent, also uses `@connectable`. A few differences in implementation (imo it's cleaner) that makes especially the parsing part more performant (uses the proper parser and caches the results). The runtime aspect should also be slightly faster (not verified) because it has less work to do.

(This is a pretty important piece that was missing.)

bigopon commented on packages/runtime/src/binding/ast.ts:
```diff
@@ -923,29 +924,33 @@ export class ForOfStatement implements IExpression {
+      return parts[0] + this.firstExpression.evaluate(flags, scope, locator) + parts[1];
     }
   }
+  public connect(flags: BindingFlags, scope: IScope, binding: IConnectableBinding): void { }
```
Is this intentionally left out?

fkleuver replied:
Yeah, it's just a noop like some other AST classes have that cant be / dont make sense to connect. It's just there to satisfy the interface. I also don't really see a point in throwing. It's just kind of there. Maybe the interface needs to change or something

bigopon replied:
I was confused at first, but it seems connecting job is delegated to the expression part, so this is fine. Probably in a polishing final touch, we may want to comment the flow how a interpolation is translated and observed

bigopon commented:
@fkleuver what is the benefit of having two dedicated classes to handle the interpolation vs the existing way? They look similar to me.

fkleuver commented:
@bigopon They only look similar to the existing way because I've simplified/normalized some of the logic from vCurrent and offloaded some responsibility to the `Interpolation` expression type. There is still a key difference in that this mechanism allows both a single and multiple interpolations to effectively be treated as a single binding from the renderer/view's perspective.

The `MultiInterpolationBinding` corresponds to vCurrent's [InterpolationBinding](https://github.com/aurelia/templating-binding/blob/master/src/interpolation-binding-expression.js#L49) and the `InterpolationBinding` corresponds to [ChildInterpolationBinding](https://github.com/aurelia/templating-binding/blob/master/src/interpolation-binding-expression.js#L116)

The existing mechanism would break (not necessarily error, but unpredictable / wrong behavior) with more than one interpolation per binding with value converts and/or binding behaviors.
The other thing (albeit small) is performance. Interpolation binding can never be two-way so it needs a bit less stuff on it.

bigopon commented:
I was under the impression that `Binding` + `Template` ast could replace the need for dedicated interpolation bindings. Maybe I was wrong. I'll have a look and do a follow up PR with more examples around this. Other than that, lgtm.

fkleuver commented:
Perhaps it can, but the subtle differences would create a need for a few extra checks. The expressions of a template (by their parsing rules) cannot contain value converters or binding behaviors. Theoretically we could introduce this and make template behave the same way as interpolation, but that wouldn't remove the need for InterpolationBinding (in fact it would create the need for a recursive one)

There are more things we could cram into single classes but that wouldn't necessarily benefit library size more than it would negatively impact performance/maintainability.
